Fix parsing logic in `proc_macro::quote`

moatom opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->
Fix https://github.com/rust-lang/rust/issues/140238

moatom committed:
Fix parsing logic in `proc_macro::quote`

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18943603379/job/54088717633) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/54088717633) [(plain)](https://github.com/rust-lang/rust/commit/02f04b0d09ef588dd8bf88e608e1ac7637c885ad/checks/54088717633/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/macros/macro-quote-test.rs stdout ----

error: auxiliary build of /checkout/tests/ui/macros/auxiliary/hello_macro.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/auxiliary/hello_macro.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/macros/macro-quote-test/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "proc-macro" "--extern" "proc_macro" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/macros/macro-quote-test/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0425]: cannot find value `t` in this scope
##[error]  --> /checkout/tests/ui/macros/auxiliary/hello_macro.rs:14:32
   |
LL |         macro_rules! m { ($$($$t:tt)*) => { $$($$t)* } }
   |                                ^ not found in this scope

error[E0425]: cannot find value `t` in this scope
##[error]  --> /checkout/tests/ui/macros/auxiliary/hello_macro.rs:14:50
   |
LL |         macro_rules! m { ($$($$t:tt)*) => { $$($$t)* } }
   |                                                  ^ not found in this scope

error[E0282]: type annotations needed
##[error]  --> /checkout/tests/ui/macros/auxiliary/hello_macro.rs:14:32
   |
LL |         macro_rules! m { ($$($$t:tt)*) => { $$($$t)* } }
   |                                ^ cannot infer type

error[E0282]: type annotations needed
##[error]  --> /checkout/tests/ui/macros/auxiliary/hello_macro.rs:14:50
   |
LL |         macro_rules! m { ($$($$t:tt)*) => { $$($$t)* } }
   |                                                  ^ cannot infer type

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0282, E0425.
For more information about an error, try `rustc --explain E0282`.
------------------------------------------

---- [ui] tests/ui/macros/macro-quote-test.rs stdout end ----
---- [ui] tests/ui/macros/same-sequence-span.rs stdout ----

error: auxiliary build of /checkout/tests/ui/macros/auxiliary/proc_macro_sequence.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/auxiliary/proc_macro_sequence.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/macros/same-sequence-span/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "proc-macro" "--extern" "proc_macro" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/macros/same-sequence-span/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0425]: cannot find value `x` in this scope
##[error]  --> /checkout/tests/ui/macros/auxiliary/proc_macro_sequence.rs:13:18
   |
LL |             (1 $$x:expr $$($$y:tt,)* $$(= $$z:tt)*) => {};
   |                  ^ not found in this scope

error[E0425]: cannot find value `y` in this scope
##[error]  --> /checkout/tests/ui/macros/auxiliary/proc_macro_sequence.rs:13:30
   |
LL |             (1 $$x:expr $$($$y:tt,)* $$(= $$z:tt)*) => {};
   |                              ^ not found in this scope

error[E0425]: cannot find value `z` in this scope
##[error]  --> /checkout/tests/ui/macros/auxiliary/proc_macro_sequence.rs:13:45
   |
LL |             (1 $$x:expr $$($$y:tt,)* $$(= $$z:tt)*) => {};
   |                                             ^ not found in this scope

warning: unused imports: `Span` and `TokenTree`
##[warning]  --> /checkout/tests/ui/macros/auxiliary/proc_macro_sequence.rs:5:25
   |
LL | use proc_macro::{quote, Span, TokenStream, TokenTree};
   |                         ^^^^               ^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` (part of `#[warn(unused)]`) on by default

error[E0282]: type annotations needed
##[error]  --> /checkout/tests/ui/macros/auxiliary/proc_macro_sequence.rs:13:30
   |
LL |             (1 $$x:expr $$($$y:tt,)* $$(= $$z:tt)*) => {};
   |                              ^ cannot infer type

error[E0282]: type annotations needed
##[error]  --> /checkout/tests/ui/macros/auxiliary/proc_macro_sequence.rs:13:45
   |
LL |             (1 $$x:expr $$($$y:tt,)* $$(= $$z:tt)*) => {};
   |                                             ^ cannot infer type

error: aborting due to 5 previous errors; 1 warning emitted

Some errors have detailed explanations: E0282, E0425.
For more information about an error, try `rustc --explain E0282`.
------------------------------------------

---- [ui] tests/ui/macros/same-sequence-span.rs stdout end ----
---- [ui] tests/ui/proc-macro/generate-dollar-ident.rs stdout ----

error: auxiliary build of /checkout/tests/ui/proc-macro/auxiliary/generate-dollar-ident.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/proc-macro/auxiliary/generate-dollar-ident.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/proc-macro/generate-dollar-ident/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "proc-macro" "--extern" "proc_macro" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/proc-macro/generate-dollar-ident/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0425]: cannot find value `var` in this scope
##[error]  --> /checkout/tests/ui/proc-macro/auxiliary/generate-dollar-ident.rs:10:24
   |
LL |         $black_hole!($$var);
   |                        ^^^ not found in this scope
   |
help: consider importing one of these functions
   |
```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18945982247/job/54097368402) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/54097368402) [(plain)](https://github.com/rust-lang/rust/commit/36c13e990088728d070eb05709d923470e1fa829/checks/54097368402/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/proc-macro/quote/basic.rs stdout ----

error: auxiliary build of /checkout/tests/ui/proc-macro/quote/auxiliary/basic.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/proc-macro/quote/auxiliary/basic.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/proc-macro/quote/basic/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "proc-macro" "--extern" "proc_macro" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/proc-macro/quote/basic/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0425]: cannot find value `j` in this scope
##[error]  --> /checkout/tests/ui/proc-macro/quote/auxiliary/basic.rs:66:35
   |
LL |     let tokens = quote!($$ j $($$ j $iter)*);
   |                                   ^ help: a unit struct with a similar name exists: `X`
...
LL | struct X;
   | --------- similarly named unit struct `X` defined here

error[E0282]: type annotations needed
##[error]  --> /checkout/tests/ui/proc-macro/quote/auxiliary/basic.rs:66:35
   |
LL |     let tokens = quote!($$ j $($$ j $iter)*);
   |                                   ^ cannot infer type

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0282, E0425.
For more information about an error, try `rustc --explain E0282`.
```

</details>

moatom commented:
@bors r+

bors commented:
@moatom: :key: Insufficient privileges: Not in reviewers

dtolnay commented on tests/ui/proc-macro/quote/auxiliary/basic.rs:
```diff
@@ -50,6 +53,23 @@ pub fn run_tests(_: TokenStream) -> TokenStream {
+fn test_dollar_dollar() {
+    let iter = ["a", "b"].into_iter();
+    let x = "X";
+    let tokens = quote!($$ x $($$ x $iter)*);
```
This case is not actually fixed. It will fail to compile if you remove the unused definition of `x`.

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -477,6 +407,381 @@ fn collect_meta_vars(content_stream: TokenStream) -> Vec<Ident> {
+        [TokenTree::Punct(t1)] if t1.as_char() == '_' => true,
+        [TokenTree::Ident(_)] => true,
+        [TokenTree::Group(_)] => true,
+        [_x] => true,
```
This is incorrectly returning true for `TokenTree::Punct('\'', Spacing::Joint)`.

```suggestion
        [TokenTree::Ident(_) | TokenTree::Group(_) | TokenTree::Literal(_)] => true,
```

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -477,6 +407,381 @@ fn collect_meta_vars(content_stream: TokenStream) -> Vec<Ident> {
+                let sep_opt: Option<TokenStream> =
+                    (!sep_cand.is_empty()).then(|| sep_cand.into_iter().collect::<TokenStream>());
+
+                expand_dollar_group_sep_star(current_contents.clone(), sep_opt)
```
```suggestion
                expand_dollar_group_sep_star(current_contents, sep_opt)
```

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -477,6 +407,381 @@ fn collect_meta_vars(content_stream: TokenStream) -> Vec<Ident> {
+                is_sep_confirmed = false;
+
+                // Move to consume the next `$( CONTENTS )`.
+                current_contents = next_group.stream().clone();
```
```suggestion
                current_contents = next_group.stream();
```

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -477,6 +407,381 @@ fn collect_meta_vars(content_stream: TokenStream) -> Vec<Ident> {
+                .to_tokens(&mut tokens);
+
+                let sep_opt: Option<TokenStream> = (!sep_cand.is_empty())
+                    .then(|| sep_cand.clone().into_iter().collect::<TokenStream>());
```
```suggestion
                    .then(|| sep_cand.into_iter().collect::<TokenStream>());
```

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -477,6 +407,381 @@ fn collect_meta_vars(content_stream: TokenStream) -> Vec<Ident> {
+    content_stream: TokenStream,
+    iter: &mut Peekable<crate::token_stream::IntoIter>,
+) -> TokenStream {
+    let mut tokens = crate::TokenStream::new();
```
```suggestion
    let mut tokens = TokenStream::new();
```

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -477,6 +407,381 @@ fn collect_meta_vars(content_stream: TokenStream) -> Vec<Ident> {
+            )
+        } else {
+            minimal_quote!(();)
+        })
```
```suggestion
        (@ if let Some(sep) = sep_opt {
            Some(minimal_quote!(
                if _i > 0 {
                    (@ quote(sep)).to_tokens(&mut ts);
                }
                _i += 1;
            ))
        } else {
            None
        })
```

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -477,6 +407,381 @@ fn collect_meta_vars(content_stream: TokenStream) -> Vec<Ident> {
+            }
+
+            // If the next `$( CONTENTS )` is found before the `*`, consume the current `$( CONTENTS )` literally.
+            // Then move to consume the next `$( CONTENTS )`.
```
I do not recognize what case this arm is trying to handle. I also was unable to come up with any `quote!` invocation where this arm would expand to something that even compiles. Please add a test that covers what is being done here.

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -284,96 +286,24 @@ pub fn quote(stream: TokenStream) -> TokenStream {
             match tree {
-                TokenTree::Group(tt) => {
+                TokenTree::Group(ref tt) => {
                     // Handles repetition by expanding `$( CONTENTS ) SEP_OPT *` to `{ REP_EXPANDED }`.
```
It can be tested by calling `quote!` from a macro_rules macro.

```rust
macro_rules! do_quote {
    ($dollar:tt $content:expr) => {
        proc_macro::quote!($dollar $content *)
    };
}

let arr = [0; 2];
eprintln!("{}", do_quote!($ f!($arr)));
```

On current nightly this nonsensically prints `f! (0i32) f! (0i32)`.

dtolnay commented on library/proc_macro/src/quote.rs:
```diff
@@ -450,20 +390,34 @@ pub fn quote(stream: TokenStream) -> TokenStream {
-            match &tree {
-                TokenTree::Punct(tt) if tt.as_char() == '$' => {
-                    if let Some(TokenTree::Ident(id)) = iter.peek() {
+        while let Some(ref tree) = iter.next() {
```
This is going to miss adjacent metavariables. For example this works:

```rust
let a = ['a'];
let b = ['b'];
eprintln!("{}", proc_macro::quote!($($a . $b)*));
```

but this does not:

```rust
eprintln!("{}", proc_macro::quote!($($a $b)*));
```

moatom replied:
[note] This was caused by an extra `iter.next();`: https://github.com/rust-lang/rust/pull/148209/commits/b2ae84c8b6a1aaa35154e695c9d34d21193dc0fe#diff-6ebaa6be050b7cea1d542579c11cde3cba0942a3417c53298001f9db02a8d06dL407

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19476089080/job/55736220609) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/55736220609) [(plain)](https://github.com/rust-lang/rust/commit/ae0617b3066333641cebecf4ef7aa08d3e284072/checks/55736220609/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling twox-hash v1.6.3
error[E0382]: borrow of moved value: `while_body`
   --> compiler/rustc_proc_macro/../../library/proc_macro/src/quote.rs:625:16
    |
595 |     let mut while_body = TokenStream::new();
    |         -------------- move occurs because `while_body` has type `TokenStream`, which does not implement the `Copy` trait
...
599 |         TokenTree::Group(Group::new(Delimiter::Brace, while_body)),
    |                                                       ---------- value moved here
...
625 |     .to_tokens(&mut while_body);
    |                ^^^^^^^^^^^^^^^ value borrowed here after move
    |
note: consider changing this parameter type in method `new` to borrow instead if owning the value isn't necessary
   --> compiler/rustc_proc_macro/../../library/proc_macro/src/lib.rs:818:46
    |
818 |     pub fn new(delimiter: Delimiter, stream: TokenStream) -> Group {
    |            --- in this method                ^^^^^^^^^^^ this parameter takes ownership of the value
help: consider cloning the value if the performance cost is acceptable
    |
599 |         TokenTree::Group(Group::new(Delimiter::Brace, while_body.clone())),
    |                                                                 ++++++++

For more information about this error, try `rustc --explain E0382`.
[RUSTC-TIMING] rustc_proc_macro test:false 0.517
error: could not compile `rustc_proc_macro` (lib) due to 1 previous error
```

</details>

moatom commented:
@dtolnay

> I've addressed your comments and left the fix commit separate for traceability.
Please let me know if it's okay to squash.

I've now rebased and squashed the commits for smoother communication.
If you still need the detailed history, please take a look at [moatom:stabilize-proc-macro-saved](https://github.com/moatom/rust/commits/stabilize-proc-macro-saved/).
